File: moqt/internal/message/subscribe_ok.go
package message

import (
	"errors"
	"io"
)

var ErrInvalidSubscribeOkMessageType = errors.New("invalid message type for SubscribeOkMessage")

// Type tags for messages sent by the publisher on the Subscribe Stream.
// The type varint is written by the caller before Encode / consumed before Decode.
const (
	MessageTypeSubscribeOk   uint64 = 0x0
	MessageTypeSubscribeEnd  uint64 = 0x1
	MessageTypeSubscribeDrop uint64 = 0x2
)

// SubscribeOkMessage confirms a subscription and resolves its absolute
// start group. It is the first message the publisher sends on the
// Subscribe Stream, once the start group is known.
//
// Wire format:
//
//	SUBSCRIBE_OK Message {
//	  Type (i) = 0x0
//	  Message Length (i)
//	  Group (i)
//	}
type SubscribeOkMessage struct {
	// Group is the absolute sequence number of the first group that will
	// be delivered (plain absolute sequence, not the +1 form of SUBSCRIBE).
	Group uint64
}

func (som SubscribeOkMessage) Len() int {
	return VarintLen(som.Group)
}

func (som SubscribeOkMessage) Encode(w io.Writer) error {
	msgLen := som.Len()
	b := make([]byte, 0, msgLen+VarintLen(uint64(msgLen)))

	b, _ = WriteMessageLength(b, uint64(msgLen))
	b, _ = WriteVarint(b, som.Group)

	_, err := w.Write(b)

	return err
}

func (som *SubscribeOkMessage) Decode(src io.Reader) error {
	num, err := ReadMessageLength(src)
	if err != nil {
		return err
	}

	if num > MaxMessageSize {
		return ErrMessageTooLarge
	}

	b := make([]byte, num)
	_, err = io.ReadFull(src, b)
	if err != nil {
		return err
	}

	num, n, err := ReadVarint(b)
	if err != nil {
		return err
	}
	som.Group = num
	b = b[n:]

	if len(b) != 0 {
		return ErrMessageTooShort
	}

	return nil
}

File: moq-web/src/internal/message/subscribe_ok.ts
import type { Reader, Writer } from "@okdaichi/golikejs/io";
import { MessageDecoder, MessageEncoder, readFull, readVarint } from "./message.ts";

export interface SubscribeOkMessageInit {
	group?: number;
}

/**
 * SUBSCRIBE_OK message (type 0x0), confirming a subscription and resolving
 * its absolute start group. The type varint is written by the caller before
 * `encode` / consumed before `decode`.
 *
 * ```text
 * SUBSCRIBE_OK Message {
 *   Type (i) = 0x0
 *   Message Length (i)
 *   Group (i)
 * }
 * ```
 */
export class SubscribeOkMessage {
	/**
	 * Absolute sequence of the first group that will be delivered
	 * (plain absolute sequence, not the +1 form used in SUBSCRIBE).
	 */
	group: number;

	constructor(init: SubscribeOkMessageInit = {}) {
		this.group = init.group ?? 0;
	}

	/**
	 * Encodes the message to the writer.
	 */
	async encode(w: Writer): Promise<Error | undefined> {
		const e = new MessageEncoder();
		e.varint(this.group);
		const [, err] = await w.write(e.frame());
		return err;
	}

	/**
	 * Decodes the message from the reader.
	 */
	async decode(r: Reader): Promise<Error | undefined> {
		const [msgLen, , err1] = await readVarint(r);
		if (err1) return err1;

		const buf = new Uint8Array(msgLen);
		const [, err2] = await readFull(r, buf);
		if (err2) return err2;

		const d = new MessageDecoder(buf);
		this.group = d.varint();

		return undefined;
	}
}
